### PR TITLE
[MJAVADOC-775] Option 'taglets/taglet/tagletpath' ignored when pointing to a JAR

### PR DESCRIPTION
**Pull request by kriegaex**
Fixes [MJAVADOC-775](https://issues.apache.org/jira/browse/MJAVADOC-775).
Fixes [MJAVADOC-783](https://issues.apache.org/jira/browse/MJAVADOC-783).

@elharo, @robjg, @jkesselm

---

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:148]: For the sake of correctness: `Locale.ROOT`

**Review by michael-o — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (kriegaex) [reply src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java]: Done, thanks.

**Comment by michael-o** [comment]
@kriegaex Thank you for the PR. I first need to understand the original reports and your change whether I can computer this reasonably. It will take a few days. Tell me please, is an IT required here in this case or is the UT enough=

**Comment by kriegaex** [comment]
@michael-o, you can find detailed information and reproducers in the two linked Jira issues.

As for ITs, taglet paths are completely untested. At least, I did not find anything there. The string "taglet" does not occur **at all** in the IT directory. I think, ITs would make sense in general. But my PR does not make IT coverage any worse than zero. As for unit tests, I added a test for the new method.

As one of the issues was closed after so-called "triage" without proper investigation, just because the two sounded similar, I was concerned that this topic would not be taken seriously by the product maintainers. So, I contributed this PR, which was not my original plan. I cannot fix all issues I encounter in all OSS products by myself, always dealing with code I have never seen before, and on top of it also improve test coverage. My own projects are already suffering more than is good for them. So please, let us not blow up the scope of this PR. But by all means, feel free to add ITs, committing on top of my changes after the merge. 🙂

**Comment by michael-o** [comment]
> @michael-o, you can find detailed information and reproducers in the two linked Jira issues.
> 
> As for ITs, taglet paths are completely untested. At least, I did not find anything there. The string "taglet" does not occur **at all** in the IT directory. I think, ITs would make sense in general. But my PR does not make IT coverage any worse than zero. As for unit tests, I added a test for the new method.
> 
> As one of the issues was closed after so-called "triage" without proper investigation, just because the two sounded similar, I was concerned that this topic would not be taken seriously by the product maintainers. So, I contributed this PR, which was not my original plan. I cannot fix all issues I encounter in all OSS products by myself, always dealing with code I have never seen before, and on top of it also improve test coverage. My own projects are already suffering more than is good for them. So please, let us not blow up the scope of this PR. But by all means, feel free to add ITs, committing on top of my changes after the merge. 🙂

If you consider both JIRA issues distinct although the fix is identical, I can reopen the closed one, no problem.

**Comment by kriegaex** [comment]
The fix is **not** identical. There are two separate commits in this PR, fixing two distinct problems. Please read the commit comments, I referenced one issue per commit. I just did not want to create two PRs, because my time is limited, and I did not want to rebase later.

**Comment by michael-o** [comment]
> The fix is **not** identical. There are two separate commits in this PR, fixing two distinct problems. Please read the commit comments, I referenced one issue per commit. I just did not want to create two PRs, because my time is limited, and I did not want to rebase later.

My bad.

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:163]: the collection of <code>String</code> directories --> the paths of the directories
are these relative or absolute?

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:164]: what does valid mean here?

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:130]: what does valid mean here?

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:122]: What does invalid mean in this context? Rewrite Javadoc to clarify

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:128]: the collection of <code>String</code> paths --> the paths
Javadoc will insert the types

**Review by elharo — CHANGES_REQUESTED** [review]
I do think we need an integration test for this, or perhaps expanded unit test coverage. Without that we don't know that taglet paths now work.

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (kriegaex) [reply src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:164]: Why are you asking questions about existing comments in your product, which I have just copied/moved and slightly adjusted? If it was good enough so far, it should be good enough now.

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (kriegaex) [reply src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:130]: Why are you asking questions about existing comments in your product, which I have just copied/moved and slightly adjusted? If it was good enough so far, it should be good enough now.

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (kriegaex) [reply src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:122]: Why are you asking questions about existing comments in your product, which I have just copied/moved and slightly adjusted? If it was good enough so far, it should be good enough now.

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (kriegaex) [reply src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:128]: Why are you asking questions about existing comments in your product, which I have just copied/moved and slightly adjusted? If it was good enough so far, it should be good enough now.

**Comment by kriegaex** [comment]
> I do think we need an integration test for this, or perhaps expanded unit test coverage. Without that we don't know that taglet paths now work.

You did not know before either. Like I said, feel free to write an IT. Do not abuse contributors to catch up with work that was not done when it was due for existing code.

**Comment by kriegaex** [comment]
> Rewrite Javadoc to clarify

@elharo, if you did not notice by inspecting my PR: I just contributes something valueable, fixing two bugs within 30 minutes of seeing this code for the first time in my life. It is meant to be a minimally invasive change, not a structural refactoring of either code or javadoc. Hence, I kept wording and structure of the existing javadocs intact. Otherwise, I would have had to rewrite the javadocs for the whole class and maybe also for the rest of the module.

FYI, I also checked whom you can ask what this javadoc means and why it was written that way: Vincent Siveton.

There is a software craftsmanship principle which is often forgotten: to separate refactoring functional changes. This PR is about functional changes and not supposed to restructure the code or its documentation more than necessary to accomodate to the fact that I made one method more flexible by and additional parameter and use that method from the previously existing one in to avoid code duplication.

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:128]: Because you copied/moved and slightly adjusted them. You also added new emthods that weren't here before.

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:130]: There was no prunePaths method here before. If you're adding a public method, clear documentation is required.

Comment (kriegaex) [reply src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:130]: Look, I understand if you are very busy and do not have enough time to triage bugs or review code. But if you cannot see the difference between a new method and a renamed, slightly extended one to convey the more generic behaviour, differentiating it by name from the original method the code was factored out from, I really cannot help you. I can live without you consenting to and merging these two bugfixes, if you want to coerce me to do your job to refactor the existing documentation, after I already did your job of fixing two old bugs. Merge it, fine. Refactor the docs after the merge, be my guest. Or don't merge it, hurting yourself and your users, also fine. I am done here.

File: src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java
Comment (kriegaex) [reply src/main/java/org/apache/maven/plugins/javadoc/JavadocUtil.java:128]: You don't get it, do you? Then have your way, I could not care less. If you cannot see that this is PR is an improvement and refuse to accept it, you are just shooting in your own foot.

**Comment by kriegaex** [comment]
> > I do think we need an integration test for this, or perhaps expanded unit test coverage. Without that we don't know that taglet paths now work.
> 
> You did not know before either. Like I said, feel free to write an IT. Do not abuse contributors to catch up with work that was not done when it was due for existing code.

@elharo, I forgot to mention a detail: When you start writing ITs and want to keep the build running on both JDK 8 and JDK 9+ (CI builds currently run on 8 and 17), enjoy implementing all sample taglets twice, once for the old Doclet API and again for the Doclet APi from JDK 9+, see [JEP 221](https://openjdk.org/jeps/221). You can have the same fun I had when helping out a bit at Xalan-Java, creating [JDK-dependent Java profiles](https://github.com/apache/xalan-java/blob/9d0fd070e88474a96713b1c74bdfefcf89e04674/pom.xml#L46-L74) adding a [JDK 8 taglet module](https://github.com/apache/xalan-java/tree/xalan-java-mvn-refactored/xalan2jtaglet) for the old API and a [JDK 9+ one](https://github.com/apache/xalan-java/tree/xalan-java-mvn-refactored/xalan2jtaglet_jdk9) for the new API. All of this only to have sample code for ITs.

Of course, this makes sense, but it is way off limits for this bugfix PR. Before anyone creates such ITs, the Maven Javadoc team should maybe first decide, if the small benefit of being able to still build (not run, only build!) the plugin on JDK 8 is worth the more than duplicate (because of extra profiles and distinction of cases) effort of creating and maintaining such ITs. After that decision, work on ITs can start in a separate PR.

**Review by michael-o — APPROVED** [review]
This is good enough for me now, will merge as soon as CI passes.

Note that I have lifted the restrictions on JAR files for two reasons:
* We don't filter with `tagletpath`, it should be consistent
* ZIP files are first class classpath elements as well, so we allow everything and the rest is the responsiblity of the developer

**Comment by jira-importer** [comment]
Resolve #1161

**Pull request merged**
